### ISSUE TITLE
Proposal: add a new element to spec to support for API Gateway mapping

### ISSUE THREAD

**Issue by paulcastro** [openwhisk/openwhisk-wskdeploy#102]
There is currently an experimental API for an API Gateway.  We should investigate support for this API inside a manifest and/or deployment.

The experimental API currently runs in the CLI and we can derive the calls from running the CLI in verbose mode.

An example of possible syntax:

```
api-experimental:
    newbot:
      basepath: /whiskbot
      path: /newbot
      verb: get
      action: newbot
```

where the action is `newbot`, and the constructed API is a GET call with path `/whiskbot/newbot`

**Comment by daisy-ycguo** [comment]
Good idea. It's a kind of mapping, mapping a REST GET call to an action. So I think it can be included in the deployment yaml file.

**Comment by lzbj** [comment]
Is it included in a manifest file, we may need the source file, so example:
api-experimental:
    newbot:
      src:/src/newbot.js
      basepath: /whiskbot
      path: /newbot
      verb: get
      action: newbot

just FYR.

**Comment by paulcastro** [comment]
I'm not sure we need the a `src` -- I think the specified `action` should map back to a defined entity in the manifest already. For example, see the whiskbot use case where `newbot` is already defined as a sequence, then we use api-experimental to define a REST API for it.

The danger here is that the underlying protocol for api-experimental is still ... experimental.  It may change on us without notice.

**Comment by lzbj** [comment]
Ok, clear. Maybe we could put it as a low priority one after the data processing use cases.

**Comment by mrutkows** [comment]
@daisy-ycguo @paulcastro can we create a proposal for the grammar for the spec. please? or do we want this issue to reflect the spec. and open another later for the work once we add it?

**Comment by mrutkows** [comment]
for reference: https://www.ibm.com/blogs/bluemix/2017/01/exposing-openwhisk-restful-apis-api-gateway/ 

Please note the OpenAPI config option as well when designing a proposal.

**Comment by daisy-ycguo** [comment]
@mrutkows I think we should create the grammar in spec before we start the implementation.

**Comment by daisy-ycguo** [comment]
I have two options for the proposal. @mrutkows @paulcastro @lzbj please review.

Option 1

The goal of API Gateway is to expose actions via RESTful urls. So the exposed RESTful urls could be properties (or sub elements) of actions. A new property ( or a sub element) is added to `action` element or other elements, e.g. `action sequence` and etc.

The property ( or sub elements) can be put in manifest file or deployment file. Both are OK. Below are two samples. The first one is using sub element. The second one is using properties.

Sample:
```
actions:
      fire:
        location: actions/fire/index.js
        inputs:
          triggerName:
            type: string
        exposedUrl:
            url: /whiskbot/slack-event
            verb: post

seqeunces:
      newbot: 
        actions: oauth/login, decrypt, newbot-setup
        exposedUrl: get /whiskbot/newbot
        exposedUrl: get /whiskbot/anothernewbot
```

Option 2

Introduce a new element to describe the mapping. I would avoid to use `api-experimental` because it's not generic.

Sample:
```
api-gateway-entries:
    newbot:
      basepath: /whiskbot
      path: /newbot
      verb: get
      action: newbot
```

I prefer the second one, although the first one looks simple from the perspective of pkg specification. The second one may be easier to expand in the future. I see the api-gateway will enhance the security to support API key and OAuth in the future.

**Comment by paulcastro** [comment]
@daisy-ycguo  these look like great options. I slightly prefer the second option also since we can treat it as a separate entity in our implementation.  We may want to add syntax to indicate if it is a trigger or action to make it more explicit:

```
api-gateway-entries:
   actions:
      newbot:
         url: /whiskbot/newbot
         verb: get
      anotherbot:
           url: .....
```

 I'd like to get comments from people on the gateway side to see if we're missing something.

**Comment by lzbj** [comment]
I also prefer the second option, one question, do we also need consider the API sequence, for example, we need put put/post/get to achieve some final results?

**Comment by daisy-ycguo** [comment]
@paulcastro Why do you want to differentiate actions and triggers in api-gateway-entries definition? Will they be treated differently in the back end?

@lzbj Do you mean `action sequence` ?I think we should take `action sequence` into consideration.

**Comment by paulcastro** [comment]
@daisy-ycguo mostly from a usability issue.  If someone is editing the manifest, they can more easily see where an action or trigger is if we group them together.  I'm not 100% sure this is necessary though.

**Comment by lzbj** [comment]
@daisy-ycguo , yes.

**Comment by lzbj** [comment]
I think this one duplicated with #201. close it, feel free to reopen if necessary.
